primary/src/grpc_server/validator.rs

```diff
@@ -2,40 +2,89 @@
 // SPDX-License-Identifier: Apache-2.0
 use std::time::Duration;
 
-use crate::{block_waiter::GetBlockResponse, BlockCommand};
+use crate::{block_waiter::GetBlockResponse, BlockCommand, BlockRemoverCommand};
 use tokio::{
-    sync::{mpsc::Sender, oneshot},
+    sync::{mpsc::channel, mpsc::Sender, oneshot},
     time::timeout,
 };
 use tonic::{Request, Response, Status};
 use types::{
-    BatchMessageProto, BlockError, CollectionRetrievalResult, GetCollectionsRequest,
+    BatchMessageProto, BlockError, CollectionRetrievalResult, Collections, Empty,
     GetCollectionsResponse, Validator,
 };
 
 #[derive(Debug)]
 pub struct NarwhalValidator {
     tx_get_block_commands: Sender<BlockCommand>,
     get_collections_timeout: Duration,
+    tx_block_removal_commands: Sender<BlockRemoverCommand>,
+    remove_collections_timeout: Duration,
 }
 
 impl NarwhalValidator {
     pub fn new(
         tx_get_block_commands: Sender<BlockCommand>,
         get_collections_timeout: Duration,
+        tx_block_removal_commands: Sender<BlockRemoverCommand>,
+        remove_collections_timeout: Duration,
     ) -> Self {
         Self {
             tx_get_block_commands,
             get_collections_timeout,
+            tx_block_removal_commands,
+            remove_collections_timeout,
         }
     }
 }
 
 #[tonic::async_trait]
 impl Validator for NarwhalValidator {
+    async fn remove_collections(
+        &self,
+        request: Request<Collections>,
+    ) -> Result<Response<Empty>, Status> {
+        let collection_ids = request.into_inner().collection_ids;
+        let remove_collections_response = if !collection_ids.is_empty() {
+            let (tx_remove_block, mut rx_remove_block) = channel(1);
+            let mut ids = vec![];
+            for collection_id in collection_ids {
+                ids.push(collection_id.try_into().map_err(|err| {
+                    Status::invalid_argument(format!("Could not serialize: {:?}", err))
+                })?);
+            }
+            self.tx_block_removal_commands
+                .send(BlockRemoverCommand::RemoveBlocks {
+                    ids,
+                    sender: tx_remove_block,
+                })
+                .await
+                .map_err(|err| Status::internal(format!("Send Error: {:?}", err)))?;
+            match timeout(self.remove_collections_timeout, rx_remove_block.recv()).await {
+                Ok(Some(result)) => match result {
+                    Ok(_) => Ok(Empty {}),
+                    Err(remove_block_error) => Err(Status::internal(format!(
+                        "Removal Error: {:?}",
+                        remove_block_error.error
+                    ))),
+                },
+                Ok(None) => Err(Status::internal(
+                    "Removal channel closed, no result has been received.",
+                )),
+                Err(_) => Err(Status::internal(
+                    "Timeout, no result has been received in time",
```

Comment:
The block remover is also responding with a `timeout` error. We might want to emit only one timeout error.

Reply:
Sorry can you clarify this? Only one timeout will return either the block remover which will get surfaced and return as part of this branch 

```
Err(remove_block_error) => Err(Status::internal(format!(
                        "Removal Error: {:?}",
                        remove_block_error.error
                    ))),
```

or the grpc request timeout which you commented on above. Whichever comes first will be returned.

Reply:
Yeah, basically I meant that the consumer will see two different cases of timeout - either from the block_remover or the endpoint it self. That might be a bit confusing . As a client I would expect a "timeout" error to be a unique one - the operation timed out . Again maybe not big a deal and something we can look into afterwards.

Reply:
Or we could probe for a `BlockRemoverError` with second field matching `BlockRemoverErrorType::Timeout` above. The pattern would look like this, IIUC:
```
Ok(Some(Err(BlockRemoverError(_,BlockRemoverErrorType::Timeout))))
```

Reply:
IIUC you are both stating that we should only surface only one type of timeout error so I added the matching Francois suggested to do so. If this not what you expected please add a comment to [Issue#234](https://github.com/MystenLabs/narwhal/issues/234) and I will update it after this PR. Thanks!

```
Err(remove_block_error)
       if remove_block_error.error == BlockRemoverErrorKind::Timeout =>
       {
              Err(Status::internal("Timeout, no result has been received in time"))
        }
```